### ISSUE TITLE
old batch_matmul much faster than new batch_matmul on GPU

### ISSUE THREAD

**Issue by benkrause** [chainer/chainer#2747]
I am often getting more than an order of magnitude better performance using the previous version of batch_matmul function than with the current batch_matmul function (which uses cupy.matmul). The current batch_matmul function also seems to scale very poorly with the batch size on my setup.

**Comment by niboshi** [comment]
Thank you for reporting.
Could you provide the code and version/commit you compared to?

**Comment by benkrause** [comment]
For the old batch_matmul, I used the version of matmul committed August 23 2016.  I copy pasted code from this github version to obtain these results. All my experiments were performed using chainer 1.23 and cuda toolkit 8. I noticed the issue is mainly with batches of larger matrices (for instance 2000x2000), and with smaller matrices the new batch_matmul is actually sometimes faster. I also noticed the relative performance seems to be different across different setups. In my code on my setup (using a gtx Titan with 6GB), 100 iterations of chainer.functions.batch_matmul(A,B) where A and B are both float32 cupy arrays of shape (5,2000,2000) takes ~6 seconds. The same benchmark for the old batch_matmul takes around 0.3 seconds. Also increasing the batch from 5 to 10 doubles the computation time for chainer.functions.batch_matmul, but only increases the time for the old batch_matmul by about 10%. 

[matmul_benchmark_new.txt](https://github.com/pfnet/chainer/files/1014892/matmul_benchmark_new.txt)
[matmul_benchmark_old.txt](https://github.com/pfnet/chainer/files/1014893/matmul_benchmark_old.txt)
